master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
+	}
+
+	return nil
+}
+
+func (s *asyncHalvingSearch) initialOperations(ctx context) ([]Operation, error) {
+	// The number of initialOperations will control the degree of parallelism
+	// of the search experiment since we guarantee that each validationComplete
+	// call will return a new train workload until we reach MaxTrials.
+
+	// We will use searcher config field if available.
+	// Otherwise we will default to a number of trials that will
+	// guarantee at least one trial at the top rung.
+	var ops []Operation
+	var maxConcurrentTrials int
+
+	if s.MaxConcurrentTrials != nil {
+		maxConcurrentTrials = min(*s.MaxConcurrentTrials, s.MaxTrials)
+	} else {
+		maxConcurrentTrials = max(
+			min(int(math.Pow(s.Divisor, float64(s.NumRungs-1))), s.MaxTrials),
+			1)
+	}
+
+	for trial := 0; trial < maxConcurrentTrials; trial++ {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+	return ops, nil
+}
+
+func (s *asyncHalvingSearch) trainCompleted(
+	ctx context, requestID RequestID, message Workload,
+) ([]Operation, error) {
+	s.totalStepsCompleted++
+	return nil, nil
+}
+
+func (s *asyncHalvingSearch) validationCompleted(
+	ctx context, requestID RequestID, message Workload, metrics ValidationMetrics,
+) ([]Operation, error) {
+	// Extract the relevant metric as a float.
+	metric, err := metrics.Metric(s.AsyncHalvingConfig.Metric)
+	if err != nil {
+		return nil, err
+	}
+	if !s.AsyncHalvingConfig.SmallerIsBetter {
+		metric *= -1
+	}
+
+	return s.promoteAsync(ctx, requestID, message, metric), nil
+}
+
+func (s *asyncHalvingSearch) promoteAsync(
+	ctx context, requestID RequestID, message Workload, metric float64,
+) []Operation {
+	// Upon a trial is finished, we should return at least one more trial to train unless
+	// the bracket of successive halving is finished.
+	rungIndex := s.trialRungs[requestID]
+	rung := s.rungs[rungIndex]
+	rung.outstandingTrials--
+	addedNewTrial := false
+
+	ops := make([]Operation, 0)
+	// If the trial has completed the top rung's validation, close the trial.
+	if rungIndex == s.AsyncHalvingConfig.NumRungs-1 {
+		s.trialsCompleted++
+		if !s.earlyExitTrials[requestID] {
```

Comment:
question: Why do we need this check now but we didn't need it in `sha.go`?

Reply:
I assumed that we shouldn't have a NewClose event for any trial in earlyExitTrial.  If we don't have this here, than we could submit an NewClose event for a trial that exited early.

Reply:
yeah -- does seem to make sense.. let's add this logic to `sha.go`?